cpp/src/replace/clamp.cu

```diff
@@ -338,6 +305,81 @@ std::unique_ptr<column> dispatch_clamp::operator()<struct_view>(column_view cons
   CUDF_FAIL("clamp for struct_view not supported");
 }
 
+template <>
+std::unique_ptr<column> dispatch_clamp::operator()<cudf::dictionary32>(
+  column_view const& input,
+  scalar const& lo,
+  scalar const& lo_replace,
+  scalar const& hi,
+  scalar const& hi_replace,
+  rmm::mr::device_memory_resource* mr,
+  cudaStream_t stream)
+{
+  // add lo_replace and hi_replace to keys
+  auto matched_column = [&] {
+    auto matched_view              = dictionary_column_view(input);
+    std::unique_ptr<column> result = nullptr;
+    auto add_scalar_key            = [&](scalar const& key, scalar const& key_replace) {
+      if (key.is_valid()) {
+        result = dictionary::detail::add_keys(
+          matched_view,
+          make_column_from_scalar(key_replace, 1, rmm::mr::get_current_device_resource(), stream)
+            ->view(),
+          mr,
+          stream);
+        matched_view = dictionary_column_view(result->view());
+      }
+    };
+    add_scalar_key(lo, lo_replace);
+    add_scalar_key(hi, hi_replace);
+    return result;
+  }();
+  auto matched_view = dictionary_column_view(matched_column->view());
```

Comment:
If both `lo` and `hi` scalars aren't valid, won't `matched_column` be null here?

Reply:
No. This function will not be called if they are both invalid. This check is done here:
https://github.com/rapidsai/cudf/blob/87af89b0c2b36784ca4c91dd1cd86d7b1161a082/cpp/src/replace/clamp.cu#L406-L409